Support parallel computation for NDT

Tao-Sheng opened the pull request:
Support parallel computation for NDT by OpenMP with different neighbor search method

Tao-Sheng committed:
✨ feat: Support parallel computation for NDT

Support parallel computation for NDT by OpenMP with different neighbor search method

Tao-Sheng committed:
🐞 fix: OpenMP Requires Signed Indices

Use signed std::ptrdiff_t instead of std::size_t

Tao-Sheng committed:
🐞 fix: Apply copilot suggestion

Apply copilot suggestion to fix some error

mvieth reviewed (commented):
You have to at least put a `omp critical` region around `computePointDerivatives(x)` and `updateDerivatives(...)` because they both use global variables.
Also maybe put NeighborSearchMethod into the NormalDistributionTransform class, to make it clear that they belong together?

mvieth committed:
Fix and improve NDT parallelization
- Added the switch to select different search methods to computeHessian() as well (would not make sense to use different search methods in computeDerivatives() and computeHessian())
- Added overloads to functions computePointDerivatives(), updateDerivatives(), and updateHessian() to be able to use local variables instead of the global variables point_jacobian_ and point_hessian_
- Change the parallelized region in computeDerivatives() by using the new function overloads, and by introducing a custom reduction (goal: compatibility with OpenMP 2.0). Synchronization overhead between threads is minimal, with only two critical regions at the end to accumulate, score, gradient, and hessian
- Tested with two larger point clouds, and there is indeed a nice speed-up when increasing the number of threads (however not a perfect 1-to-1 speedup because there are still other functions in NDT that are not parallelized)

mvieth committed:
Fixes per copilot

mvieth committed:
Fix formatting

mvieth committed:
Also add DIRECT27
